### PR TITLE
Use build.sh for Ubuntu inner-loop testing

### PR DESCRIPTION
**Pull request by mellinoe**
This changes the CI definition to use `build.sh` for the inner-loop testing of Ubuntu. This should let us catch build-breaking issues for x-plat builds much easier and sooner than before.

NOTE: We are using a tool called "Packer" that Matt wrote to archive the build results. Right now it is a .NET Framework application, but we can probably convert it very easily to run on .NET Core (from the tool runtime, probably) to archive x-plat build results. For now, I am just skipping that part of the definition until we go and do that work. The current (x-plat) inner-loop definitions aren't archiving this stuff anyways, so it's not really any worse than before.

I'll need some help figuring out how to test this, as I haven't made this sort of CI change before.

@joperezr , @Priya91 , @ellismg , @weshaggard

### CODE REVIEW

**Comment by mellinoe** [comment]
@dotnet-bot test ci please

**Comment by mellinoe** [comment]
@dotnet-bot test ci please

File: netci.groovy [hunk 1]
```diff
@@ -658,35 +658,52 @@ branchList.each { branchName ->
...
             Utilities.standardJobSetup(newJob, project, isPR, getFullBranchName(branchName))
             // Add the unit test results
             Utilities.addXUnitDotNETResults(newJob, 'bin/tests/**/testResults.xml')
+            def archiveContents = ''
+            if (os == 'Windows 10' || os == 'Windows 7' || os == 'Windows_NT') {
+                archiveContents = "bin/build.pack,bin/${osGroup}.AnyCPU.Debug/**,bin/ref/**,bin/packages/**,msbuild.log"
+            }
+            else {
+                // Packer.exe is a .NET Framework application. When we can use it from the tool-runtime, we can archive the ".pack" file here.
+                archiveContents = "bin/${osGroup}.AnyCPU.Debug/**,bin/ref/**,bin/packages/**,msbuild.log"
```
Comment (Priya91) [inline netci.groovy]: How about bin/AnyOS.AnyCPU ?

File: netci.groovy [hunk 2]
```diff
@@ -658,35 +658,52 @@ branchList.each { branchName ->
...
-            Utilities.setMachineAffinity(newJob, osGroup, 'latest-or-auto')
+            // Set the affinity.
+            Utilities.setMachineAffinity(newJob, os, 'latest-or-auto')
             // Set up standard options.
             Utilities.standardJobSetup(newJob, project, isPR, getFullBranchName(branchName))
             // Add the unit test results
             Utilities.addXUnitDotNETResults(newJob, 'bin/tests/**/testResults.xml')
+            def archiveContents = ''
+            if (os == 'Windows 10' || os == 'Windows 7' || os == 'Windows_NT') {
+                archiveContents = "bin/build.pack,bin/${osGroup}.AnyCPU.Debug/**,bin/ref/**,bin/packages/**,msbuild.log"
```
Comment (Priya91) [inline netci.groovy]: instead of debug, this should be ```bin/${osGroup}.AnyCPU.${configurationGroup}``` ..

File: netci.groovy [hunk 3]
```diff
@@ -658,35 +658,52 @@ branchList.each { branchName ->
     }
 }
 
-// Generate the build and test versions for Windows_NT.  When full build/run is supported on a platform, those platforms
+// Generate the build and test versions for Windows_NT and Ubuntu.  When full build/run is supported on a platform, those platforms
 // could be removed from above and then added in below.
-def supportedFullCyclePlatforms = ['Windows_NT']
+def supportedFullCyclePlatforms = ['Windows_NT', 'Ubuntu']
```
Comment (Priya91) [inline netci.groovy]: Can you rename this to Ubuntu14.04 ..

File: netci.groovy (hunk 2 above)
Comment (mellinoe) [inline netci.groovy]: Yeah, @ellismg noticed this too. I think we can fix this separately, since I'm just matching what Windows is doing here. We should fix it in all places it occurs in the file.

File: netci.groovy [hunk 4]
```diff
@@ -658,35 +658,52 @@ branchList.each { branchName ->
...
-                    batchFile("C:\\Packer\\Packer.exe .\\bin\\build.pack .\\bin")
+                    if (os == 'Windows 10' || os == 'Windows 7' || os == 'Windows_NT') {
+                        batchFile("call \"C:\\Program Files (x86)\\Microsoft Visual Studio 14.0\\VC\\vcvarsall.bat\" x86 && build.cmd /p:ConfigurationGroup=${configurationGroup} /p:OSGroup=${osGroup}")
+                        batchFile("C:\\Packer\\Packer.exe .\\bin\\build.pack .\\bin")
+                    }
+                    else if (os == 'OSX') {
+                        shell("HOME=\$WORKSPACE/tempHome ./build.sh /p:ConfigurationGroup=${configurationGroup} /p:TestWithLocalLibraries=true")
+                    }
+                    else {
+                        shell("sudo HOME=\$WORKSPACE/tempHome ./build.sh /p:ConfigurationGroup=${configurationGroup} /p:TestWithLocalLibraries=true")
```
Comment (Priya91) [inline netci.groovy]: why passing ```TestWithLocalLibraries``` for innerloop runs? Same for OSX above as well..

Comment (ellismg) [inline netci.groovy]: We also should not be running this stuff with sudo.

Comment (mellinoe) [inline netci.groovy]: TestWithLocalLibraries is sort-of required for lots of projects in x-plat builds. If the native shims are changed in the PR then this needs to be passed, because otherwise we will pull in old shims from NuGet packages. We should be able to have more fine-grained control over what "local libraries" we need to copy, but this is the best we have right now, I think.

Technically, this isn't strictly an x-plat thing, but Windows doesn't have native components that change very often, so we don't hit the issue there.

Comment (mellinoe) [inline netci.groovy]: Removing sudo, copy-paste error from outer-loop...

File: netci.groovy (hunk 3 above)
Comment (mellinoe) [inline netci.groovy]: Can I just rename them in this file or are they referenced from somewhere else?

File: netci.groovy (hunk 2 above)
Comment (ellismg) [inline netci.groovy]: You could consider using an interpolated string with checks if `OSGroup=Windows` and adds bin/build.pack or not (may remove some duplication, your all).

File: netci.groovy (hunk 4 above)
Comment (Priya91) [inline netci.groovy]: TestWithLocalLibraries will replace all the locally built bits with the package versions, so we will no longer be testing packages on Ubuntu.. And yeah, as @ellismg notes, this should not be sudo as well.

File: netci.groovy (hunk 3 above)
Comment (mellinoe) [inline netci.groovy]: For example, we pass this through to "setMachineAffinity" in some cases, and I'm not sure what that does exactly. It might be expecting "Ubuntu" rather than "Ubuntu14.04"

File: netci.groovy [hunk 5]
```diff
@@ -658,35 +658,52 @@ branchList.each { branchName ->
...
-            def newJobName = "${osGroup.toLowerCase()}_${configurationGroup.toLowerCase()}"
+            def newJobName = "${os.toLowerCase()}_${configurationGroup.toLowerCase()}"
 
             def newJob = job(getJobName(Utilities.getFullJobName(project, newJobName, isPR), branchName)) {
                 steps {
-                    batchFile("call \"C:\\Program Files (x86)\\Microsoft Visual Studio 14.0\\VC\\vcvarsall.bat\" x86 && build.cmd /p:ConfigurationGroup=${configurationGroup} /p:OSGroup=${osGroup}")
-                    batchFile("C:\\Packer\\Packer.exe .\\bin\\build.pack .\\bin")
+                    if (os == 'Windows 10' || os == 'Windows 7' || os == 'Windows_NT') {
+                        batchFile("call \"C:\\Program Files (x86)\\Microsoft Visual Studio 14.0\\VC\\vcvarsall.bat\" x86 && build.cmd /p:ConfigurationGroup=${configurationGroup} /p:OSGroup=${osGroup}")
+                        batchFile("C:\\Packer\\Packer.exe .\\bin\\build.pack .\\bin")
```
Comment (weshaggard) [inline netci.groovy]: What do we end up doing with the stuff we are packing? If it is nothing we should probably just remove it all together.

File: netci.groovy (hunk 3 above)
Comment (Priya91) [inline netci.groovy]: Renaming here will not affect anything else.. The backend support for Ubuntu14.04 is already [here](https://github.com/dotnet/dotnet-ci/blob/master/jobs/generation/Utilities.groovy#L98)

File: netci.groovy [hunk 6]
```diff
@@ -658,35 +658,52 @@ branchList.each { branchName ->
...
+            def osGroup = osGroupMap[os]
             def isPR = (branchName == 'pr') 
-            def newJobName = "${osGroup.toLowerCase()}_${configurationGroup.toLowerCase()}"
+            def newJobName = "${os.toLowerCase()}_${configurationGroup.toLowerCase()}"
 
             def newJob = job(getJobName(Utilities.getFullJobName(project, newJobName, isPR), branchName)) {
                 steps {
-                    batchFile("call \"C:\\Program Files (x86)\\Microsoft Visual Studio 14.0\\VC\\vcvarsall.bat\" x86 && build.cmd /p:ConfigurationGroup=${configurationGroup} /p:OSGroup=${osGroup}")
-                    batchFile("C:\\Packer\\Packer.exe .\\bin\\build.pack .\\bin")
+                    if (os == 'Windows 10' || os == 'Windows 7' || os == 'Windows_NT') {
```
Comment (weshaggard) [inline netci.groovy]: Should we just have two looks? One for "cmd" and one for "sh" support instead of doing these OS version checkes in multiple locations?

File: netci.groovy (hunk 4 above)
Comment (ellismg) [inline netci.groovy]: > TestWithLocalLibraries is sort-of required for lots of projects in x-plat builds. If the native shims are changed in the PR then this needs to be passed, because otherwise we will pull in old shims from NuGet packages.

We should figure out what the right long term thing is here.  Perhaps projects can define what shims they want to consume in an ItemGroup and `publishtests.targets` can try to copy them over when it is building the layout?

That probably requires buildtools work, so if we think the right thing to do is get this landed as is so we stop breaking build.sh, I'm not going to push back.  I do think that we should file a bug however.

File: netci.groovy (hunk 3 above)
Comment (mellinoe) [inline netci.groovy]: nvm, this isn't consistent, we use Ubuntu and Ubuntu14.04 both...

File: netci.groovy (hunk 4 above)
Comment (Priya91) [inline netci.groovy]: Yeah, i filed a tracking bug for this already #7146

Comment (weshaggard) [inline netci.groovy]: TestWithLocalLibraries today is a big hack and is essentially hardwired to only support some of our CoreCLR scenarios but quickly blows up for any of our other test scenarios like desktop/net native, and so TestWithLocalLibraries will be getting deleted eventually. The long term direction is to restore against the packages that were built as part of the build so that we can us the nuget logic to select the correct assets and not try to duplicate that logic in the publishtests.targets like we are doing today.

File: netci.groovy (hunk 3 above)
Comment (Priya91) [inline netci.groovy]: Ubuntu (means Ubuntu14.04) was used in the first version of this file, when we didnt have different versions of Ubuntu. Going forward with writing new code, we should stick to specific versions.

File: netci.groovy (hunk 5 above)
Comment (Priya91) [inline netci.groovy]: the packed stuff is only to ensure we dont copy duplicates of same file under different output dirs like tests directory.. But the archival for innerloop runs is to get output of build for local repro, i dont think we use it in other jobs like in the unix innerloop runs.

File: netci.groovy (hunk 3 above)
Comment (mellinoe) [inline netci.groovy]: I've cleaned up the whole file, it was very inconsistent.

File: netci.groovy [hunk 7]
```diff
@@ -658,35 +658,49 @@ branchList.each { branchName ->
...
-            // Set the affinity.  All of these run on Windows currently.
-            Utilities.setMachineAffinity(newJob, osGroup, 'latest-or-auto')
+            // Set the affinity.
+            Utilities.setMachineAffinity(newJob, os, 'latest-or-auto')
             // Set up standard options.
             Utilities.standardJobSetup(newJob, project, isPR, getFullBranchName(branchName))
             // Add the unit test results
             Utilities.addXUnitDotNETResults(newJob, 'bin/tests/**/testResults.xml')
+            def archiveContents = "bin/${osGroup}.AnyCPU.Debug/**,bin/ref/**,bin/packages/**,msbuild.log"
+            if (os.contains('Windows')) {
```
Comment (Priya91) [inline netci.groovy]: the archiveContents should now include bin/AnyOS.AnyCPU.${configurationGroup}.. This archival made sense in the initial dir structure, but with the .builds file, we drop platform agnostic libs to AnyOS output location.

File: netci.groovy (hunk 5 above)
Comment (mellinoe) [inline netci.groovy]: Yes, it is for use in getting a local repro / more information. The x-plat inner-loop runs don't currently archive anything as far as I can tell.

File: netci.groovy (hunk 6 above)
Comment (mellinoe) [inline netci.groovy]: Do you mean we should add a helper to check whether .cmd or .sh is supported? That seems reasonable

File: netci.groovy (hunk 7 above)
Comment (mellinoe) [inline netci.groovy]: I'd prefer to do it as a separate change, it's not really related to this per-se, I'm just touching the same area.

**Comment by mellinoe** [comment]
@dotnet-bot test ci please

File: netci.groovy (hunk 4 above)
Comment (markwilkie) [inline netci.groovy]: @maririos - could you take a peek at this work and perhaps make a recommendation regarding the right next steps?  I doubt it's reasonable to fix everything all at once, but I'd like to understand a concrete plan sooner than later.  You'll need to understand better from @ellismg and @mellinoe better what's needed, but I think that between the work that @chcosta and @jhendrixMSFT have done, there should be enough there to at least understand what needs to be done and bugs/issues filed.  This should go a LONG ways towards convergence...  :)

**Comment by ellismg** [comment]
BTW, It's really awesome to see this stuff coming on line.  I know getting all this stuff working has been a group effort from a bunch of folks across the team, so it's great to see us at the point where we can finally try to lock in our wins.  Congrats all around.

**Comment by mellinoe** [comment]
@dotnet-bot test ci please

**Comment by mellinoe** [comment]
Had to rebase, this last version should be good to go unless I made some syntax error.
@dotnet-bot test ci please

File: netci.groovy [hunk 8]
```diff
@@ -528,7 +528,7 @@ branchList.each { branchName ->
             Utilities.addArchival(newBuildJob, "bin/build.pack,bin/osGroup.AnyCPU.${configurationGroup}/**,bin/ref/**,bin/packages/**,msbuild.log")
 
             // Use Server GC for Ubuntu/OSX Debug PR build & test
-            def useServerGC = ((os == 'Ubuntu' || os == 'OSX') && configurationGroup == 'Release' && isPR)
+            def useServerGC = ((os == 'Ubuntu15.10' || os == 'OSX') && configurationGroup == 'Release' && isPR)
```
Comment (ellismg) [inline netci.groovy]: This isn't exactly what we had before, FWIW.

File: netci.groovy [hunk 9]
```diff
@@ -585,7 +585,7 @@ branchList.each { branchName ->
             if (isPR) {
                 // Set PR trigger.
                 // Set of OS's that work currently. 
-                if (os in ['OSX', 'Ubuntu', 'OpenSUSE13.2', 'CentOS7.1']) {
+                if (os in ['OSX', 'Ubuntu15.10', 'OpenSUSE13.2', 'CentOS7.1']) {
```
Comment (ellismg) [inline netci.groovy]: I don't think we want to do Ubuntu 15.10 per PR, the coverage from 14.04 + Centos should be sufficient?

Comment (mellinoe) [inline netci.groovy]: Yeah, that's probably fair. I'll remove this one.

File: netci.groovy (hunk 8 above)
Comment (mellinoe) [inline netci.groovy]: Yeah, that's true. Since we're not using server GC in the "regular build-test run", I figured I'd switch to using it in this run instead. The 14.04/15.10 runs should be very similar regardless.

File: netci.groovy (hunk 9 above)
Comment (ellismg) [inline netci.groovy]: Sounds reasonable.

**Comment by mellinoe** [comment]
@dotnet-bot test ci please

**Comment by mellinoe** [comment]
@dotnet-bot test ci please

File: netci.groovy (hunk 6 above)
Comment (weshaggard) [inline netci.groovy]: Either a helper or just two .foreach loops over two lists one list for the .cmd OS's and one for the .sh OS's.

**Comment by ellismg** [comment]
@mellinoe I realized a bigger problem we are going to have here.

Today, the PR job is shared between `rc2` and `master`.  But we can't use `build.sh` in RC2, so that's going to be a big problem.  Once this is merged, PRs to the rc2 branch will probably start failing their Ubuntu legs.

In addition, the way this is defined, we'll also use `build.sh` for the jobs we generate for the RC2 branches, which also seems wrong.  That you could work around by doing things conditionally based on the branch name.

I expect we are going to have to take a few more fixes for `rc2` so I would really like us to not screw up the build definitions so we can use CI to validate stuff (they are already somewhat horked due to my refactoring of how run-test.sh works, and I need to fix that, which is how I realize we'll have a problem here)

At the same time, I want to get some coverage in.  How would you feel about setting up a per merge job in master that runs build.sh based tests, at least until we figure out how to make this play nicely with RC2?

**Comment by weshaggard** [comment]
>Today, the PR job is shared between rc2 and master. But we can't use build.sh in RC2, so that's going to be a big problem. Once this is merged, PRs to the rc2 branch will probably start failing their Ubuntu legs.

@ellismg we aren't using the netci.groovy script from the particular branch? That seems busted, imagine when we do servicing off one of these release branches it would be nice if we had CI that match the toolset for the branch at that time. Do you know what we need to do to fix that?

**Comment by ellismg** [comment]
> @ellismg we aren't using the netci.groovy script from the particular branch? That seems busted, imagine when we do servicing off one of these release branches it would be nice if we had CI that match the toolset for the branch at that time. Do you know what we need to do to fix that?

Looking at the generated jobs it does not look like we are.  You can also see how #7728 is failing due to this issue.  @mmitche would know more about if we are able to generate jobs per branch.  I agree that we need to figure out a way to lock CI definitions for a branch for servicing reasons, and we also need PR definitions per release branch that don't change.  I just don't know how exactly we go about doing that.

**Comment by ellismg** [comment]
Another thing I noticed.  Because we moved to Ubuntu14.04 instead of Ubuntu, some of the job names are now different, which means we'll need to update README.md or we'll be pointing at stale job statuses.

**Comment by stephentoub** [comment]
run-test.sh includes some configuration that I don't see in build.sh, e.g.
https://github.com/dotnet/corefx/blob/master/run-test.sh#L355-L361
but maybe I'm just missing it?

**Comment by mmitche** [comment]
Yes, we can generate CI definitions per branch (a lot of projects already do, though corefx has not yet made this transition).  If it's necessary I can make it tomorrow.

**Comment by mellinoe** [comment]
Per-branch CI definitions sounds like what we need. I'd rather just wait for us to turn that on and check this in more-or-less unchanged than do some temporary work before tomorrow. I'll be watching the builds between today and tomorrow anyways, so I doubt anything will regress.

File: netci.groovy (hunk 4 above)
Comment (maririos) [inline netci.groovy]: Sure! I'll talk to them

**Comment by mmitche** [comment]
Okay I can work on that this afternoon.

**Comment by weshaggard** [comment]
@mmitche yes we should make the CI builds branch specific for corefx. Thanks for tackling that.

**Comment by mmitche** [comment]
@weshaggard @mellinoe @stephentoub 

Alright, I'm working on the one in master right now.  I need to know whether the netci.groovy file in the rc2 branch is the one we want for RC2 CI.

**Comment by mellinoe** [comment]
> I need to know whether the netci.groovy file in the rc2 branch is the one we want for RC2 CI.

I think the one in master is the one we want in RC2 right now. We should port it over if it hasn't already been. @ellismg

**Comment by ellismg** [comment]
> I think the one in master is the one we want in RC2 right now. We should port it over if it hasn't already been. @ellismg

I looked at the differences between the netci.groovy in master and in rc2.  From a spot check it looks like if it's easier, bringing the changes to rc2 would be fine.  If it's easier to just keep what we have in rc2 (maybe with a small diff to update the job names) that would be fine as well.  The only differences is we have a few additional code coverage jobs defined in master and we've removed FreeBSD testing.

If they are both easy, do whatever will produce the smaller diff (which I am guessing would be keeping what we have with minor changes to the job names).

**Comment by mmitche** [comment]
@ellismg Next question...should the coreclr that we pull be the from the same branch (name) as the corefx branch?  I'm assuming yes.

**Comment by ellismg** [comment]
Yes.

**Comment by mellinoe** [comment]
@dotnet-bot test ci please

@ellismg , @joperezr , @Priya91 I rebased this on top of the per-branch-CI work, could you guys take a look and see if it still looks correct?

File: netci.groovy [hunk 10]
```diff
@@ -585,34 +585,45 @@ def innerLoopNonWindowsOSs = ['Ubuntu', 'Ubuntu15.10', 'Debian8.2', 'OSX', 'Cent
...
 
-            // Set the affinity.  All of these run on Windows currently.
-            Utilities.setMachineAffinity(newJob, osGroup, 'latest-or-auto')
+            // Set the affinity.
+            Utilities.setMachineAffinity(newJob, os, 'latest-or-auto')
             // Set up standard options.
             Utilities.standardJobSetup(newJob, project, isPR, "*/${branch}")
             // Add the unit test results
             Utilities.addXUnitDotNETResults(newJob, 'bin/tests/**/testResults.xml')
+            def archiveContents = "bin/${osGroup}.AnyCPU.Debug/**,bin/ref/**,bin/packages/**,msbuild.log"
```
Comment (mmitche) [inline netci.groovy]: Will this include the tests on unix?  How big is this if the files don't end up hard-linked?

File: netci.groovy (hunk 5 above)
Comment (mmitche) [inline netci.groovy]: It was only used for archival so that unix inner loop runs so that we didn't copy 40GB of data back to the server (there is a lot of duplication).

File: netci.groovy (hunk 10 above)
Comment (mellinoe) [inline netci.groovy]: This is the same set of folders as Windows, so it includes the test compilation results (*Tests.dll), but it doesn't include the actual test runtime layouts (bin/tests/<OS>/*). Right now this is fairly small (~100MB uncompressed), but that's because it doesn't actually include a lot of the assemblies (i.e. it doesn't include AnyOS). But it matches the Windows set.

File: netci.groovy (hunk 5 above)
Comment (mellinoe) [inline netci.groovy]: Could you rephrase that? How did we use these archives for the unix inner loop runs? I thought these archives were just for our own use for repro's. Don't the inner-loop runs define their own process for building on Windows and transferring?

File: netci.groovy (hunk 10 above)
Comment (mmitche) [inline netci.groovy]: Okay cool that works.  With the layouts but without the hard linking the size is huge.  We can probalby just remove the packer use here then and not archive the tests.  The number of people who have asked for the unpacker has been pretty minimal, which leads me to believe people are just reproing the old-fashioned way.

File: netci.groovy (hunk 5 above)
Comment (mmitche) [inline netci.groovy]: The way that the inner loop runs work is that it builds on Windows (using the right OSGroup and runtime id), packs up the layouts, copies them over to the target OS, and the runs there.

So actually now looking at this, I believe you want to still pack for anythign that isn't going to run build.sh on the target OS.

Comment (mellinoe) [inline netci.groovy]: I'm aware of that, but this is for the Windows-standalone inner-loop build. The runs that build on Windows but run somewhere else are defined on line ~495. We are archiving the standalone Windows inner-loop results here, I've just extended it to the standalone Ubuntu run.

Comment (mmitche) [inline netci.groovy]: i think it may have been an oversight.  We used to use the /p:OSGroup=Windows_NT build for all of the cross runs.  We don't any longer but the packer use is still there.

Comment (mellinoe) [inline netci.groovy]: Let's fix this one separately if we think it's an issue. I didn't actually change this part of the definition, I just wrapped it in a Windows-specific check.

Comment (mmitche) [inline netci.groovy]: Sounds fine to me.

**Comment by mellinoe** [comment]
The definitions from "test ci please" look good to me. I'm going to merge this and I will watch the next inner loop run that happens.

**Comment by joperezr** [comment]
Looks like [this build](http://dotnet-ci.cloudapp.net/job/dotnet_corefx/job/master/job/ubuntu14.04_debug_prtest/4/consoleFull) succeeded using build.sh :+1: Great work Eric! :smile:
